### ISSUE TITLE
Remove duplicated translation entries for languages without grammatical gender

### ISSUE THREAD

**Issue by Qrox** [CleverRaven/Cataclysm-DDA#51421]
# Is your feature request related to a problem? Please describe.
`translate_gendered_line` is currently used in NPC dialogue to support grammatical genders. It seems the function supports a grammatical gender list that can be specified for a language and selects the translation context based on available genders. However the string extraction script outputs all combinations of grammatical genders to the translation files regardless of this list, making the translation files cluttered with duplicate entries for languages without grammatical gender. Therefore the same translation need to be applied to multiple entries on Transifex creating unnecessary workload for translators.

# Describe the solution you'd like
I suppose we can remove this duplication by outputing different `.po` files according to the grammatical gender list of a language. A solution that comes to my mind is to create language-specific `xxx_from_json_LANG.py` for language LANG, based on the grammatical genders of the language, and create the `.po` file by calling `xgettext` with the language-specific files along with other non-language-specific files.

# Describe alternatives you've considered
N/A

# Additional context
Pinging @jbytheway because you are the author of related code. Also pinging @BrettDong.

**Comment by BrettDong** [comment]
I'm afraid we can only have one translation template for all languages.

**Comment by Qrox** [comment]
Is that a limitation of Transifex? Because it seems to me that it's possible to create separate templates and merge them with respective `.po` files.

**Comment by BrettDong** [comment]
In Transifex workflow a translation template corresponds to a resource (for example `master-cataclysm-dda` resource tracks _the_ translation template of master branch on GitHub), and this resource is to be translated by different language teams, then translations of this resource is then pulled back to master branch on GitHub. It didn't expect translation template could be different for different languages.

**Comment by jbytheway** [comment]
I did realise this would be an issue when I designed the feature, but because of these limitations I couldn't think of a way around it. If someone does that would be great.

Perhaps we cannot commit the strings entirely for languages that don't need them, but maybe we could auto-insert a default translation, like "this string not required for this language" before sending to Transifex so that they don't show up as needing translation there?

**Comment by BrettDong** [comment]
The translation template (the list of languages to be translated) has to be the same across different languages.

**Comment by jbytheway** [comment]
Yes, but can we automatically add translations for some at the same time the template is uploaded to Transifex?

**Comment by BrettDong** [comment]
So the workflow will then be

1. Generate _the_ translation template (`./lang/update_pot.sh`)
2. Push template to Transifex (`tx push -s`)
3. Wait for Transifex server to process the new template and update their internal data representations (I don't know how long we should wait)
4. Pull updated translations from Transifex (`tx pull -a -f`)
5. Have some script to manipulate translations data in `./lang/po/<lang>.po` to insert default translations
6. Push the edited translations back to Transifex server (`tx push -t`)

And the problems I can see now are
1. The technical complexity is getting too high. The maintenance cost of the "push template" workflow is much higher.
2. Step (5) is potentially dangerous. We have to ensure _continuously over time_ that our script (1) does the job expected; and (2) does not mess up other parts of translation data. Otherwise there is a risk of losing translation data for up to a week (restore back-up from GitHub repo).
3. It's quite often that Transifex endpoint returns HTTP 502 or HTTP 504 error when pushing or pulling data. The only thing we can do is retry until success. Pulling and pushing translations could take up to tens of minutes. If there are translators translating at the same time, there is a risk that step (6) could potentially erase their translations after step (4).

My thoughts on grammatical gender is that the cost of running it with all kinds of hacks on a translation system (`gettext`) that does not support it is too high, and it's a cost to be borne by both developers and translators. Switching to a new translation system with built-in support for grammatical gender helps in the long term, but the migration is also very costly.

**Comment by andrei8l** [comment]
> My thoughts on grammatical gender is that the cost of running it with all kinds of hacks on a translation system (`gettext`) that does not support it is too high, and it's a cost to be borne by both developers and translators. Switching to a new translation system with built-in support for grammatical gender helps in the long term, but the migration is also very costly.

The cost on translators should be the same whether the system supports grammatical gender or you hack it in with contexts (for example). It's just another msgid and some people *do* really care about grammatical correctness.

My quick [demo](https://github.com/andrei8l/Cataclysm-DDA/commit/8878650a9195b2670892569cc43809ecc96fdc60)  for adding grammatical gender support is just ~30 loc so it doesn't seem like a big technical burden either. The helpers needed for passing the avatar's gender to the translation library would be more burdensome, and those are needed regardless of the system used.

**Comment by BrettDong** [comment]
One of the costs for translators is they have to re-translate the same sentence for all combinations of genders for languages that do not have grammatical gender.

This is not a problem for plural forms because different plural forms of a message are still treated as a single entry in both PO and MO, where the singular form is the message key, and there is a well-defined function that maps an integer `n` to the index of which plural form to be used. (For example, mapping function for Russian is `(n%10==1 && n%100!=11 ? 0 : n%10>=2 && n%10<=4 && (n%100<12 || n%100>14) ? 1 : n%10==0 || (n%10>=5 && n%10<=9) || (n%100>=11 && n%100<=14)? 2 : 3)`)

As an example, `battery / batteries` will only appear once in the translation template:
```po
#: lang/json/AMMO_from_json.py
msgid "battery"
msgid_plural "batteries"
```

and translators can freely decide how many plural forms to translate for this entry, depending on their language.

Japanese translators need to translate the entry **once** in one form:
```po
msgstr[0] "電池"
```

Russian translators translates the entry **once** into four forms:
```po
msgstr[0] "батарейка"
msgstr[1] "батарейки"
msgstr[2] "батареек"
msgstr[3] "батарейки"
```

---

However, for grammatical gender, the message is replicated twice in the translation template, and translators in all language **have to** translate them twice, regardless of the nature of the language:

```po
msgctxt "memorial_male"
msgid "Put out the fire."
msgstr "消火しました。"

msgctxt "memorial_female"
msgid "Put out the fire."
msgstr "消火しました。"
```

If it's a long sentence with multiple subjects or objects, like
```po
msgid "<npcname> puts their %1$s in the %2$s's %3$s."
```

Are we going to duplicate the message 4 times in the translation template?

**Comment by BrettDong** [comment]
Another problem of migrating existing messages to supporting grammatical gender, regardless of using context or hacking `msgid`, is that we will lost all existing translations, edit history of the entry on Transifex, and Translation Memory feature on Transifex.

An original message `A` will then become two entries `A_male` and `A_female`, and everyone will have to retranslate `A_male` and `A_female`. This process can be partially automated, but some effort is needed to make the tool.

Edit history of entry `A` is then lost. Edit history of `A_male` and `A_female` will not include history of `A`, and there is no way to solve this problem. Translation Memory feature will also be impaired, meaning that if message `B` was translated at some point in the past, when `B_male` and `B_female` is added to the template in the future, Transifex will not be able to automatically fill in translations already been done in the past.

Above two examples are saying that changing the message key, either `msgid` or context, will result in a split in translation history and all the related features on Transifex, which is a significant cost should be wary of.

**Comment by andrei8l** [comment]
> However, for grammatical gender, the message is replicated twice in the translation template, and translators in all language have to translate them twice, regardless of the nature of the language:

Assuming you'd be using the prefix approach, you can handle this relatively  cheaply if you process the IDs while loading the document and store the messages in gender-specific bins, keyed by IDs stripped of prefix. Then pick the first available bin (or some mapped bin) if the requested gender bin does not contain the ID. Then translators can ignore IDs for grammatical genders that do not exist in their target language.

You're right about edit history of course; I didn't think of that.

**Comment by BrettDong** [comment]
If there is a low-cost solution to support grammatical gender, I'm more than happy to bring it in. However, up till now, I don't see any approach that doesn't come with huge pain.

**Comment by jbytheway** [comment]
> So the workflow will then be
> 
>     1. Generate _the_ translation template (`./lang/update_pot.sh`)
> 
>     2. Push template to Transifex (`tx push -s`)
> 
>     3. Wait for Transifex server to process the new template and update their internal data representations (I don't know how long we should wait)
> 
>     4. Pull updated translations from Transifex (`tx pull -a -f`)
> 
>     5. Have some script to manipulate translations data in `./lang/po/<lang>.po` to insert default translations
> 
>     6. Push the edited translations back to Transifex server (`tx push -t`)

OK.  I was hoping it could be done at the same time as pushing the template.  If that's impossible, then I agree that this approach is probably a non-starter.

> However, for grammatical gender, the message is replicated twice in the translation template, and translators in all language **have to** translate them twice, regardless of the nature of the language:

Minor niggle: with the existing solution they don't have to translate them twice, they only have to translate the ones that matter for their language and can ignore the others or use placeholder translations.  However, I realize that this is almost the same burden, so it doesn't help much.

More importantly, this only applies for strings explicitly tagged as needing grammatical gender.  That's only a tiny fraction of all the strings in the game (right now, it's just one); do you really fear that this number will grow so much that this becomes a serious concern?

**Comment by Qrox** [comment]
It seems you can [import partially translated files](https://docs.transifex.com/translation/importing-translations) to Transifex and keep the translation that is not present in the partially translated files. Would it be possible to upload the template first, then generate a `.po` file with only placeholder translations, and import this partially translated file via command line?
